### PR TITLE
bpo-44050: Extension modules can share state when they don't support sub-interpreters.

### PR DESCRIPTION
**Pull request by shihai1991**
https://bugs.python.org/issue44050

Automerge-Triggered-By: GH:encukou

### CODE REVIEW

**Comment by encukou** [comment]
I don't understand this PR. It allows Python objects to be shared between interpreters, doesn't it?

**Comment by shihai1991** [comment]
> I don't understand this PR. It allows Python objects to be shared between interpreters, doesn't it?

Hi, petr. The details in this PR: https://github.com/python/cpython/commit/82c83bd907409c287a5bd0d0f4598f2c0538f34d
Only the main interp can update the extension and def.m_base.m_copy now. So I use the `def->m_slots` to identify the extension module created from `PyModule_Create()` or not. If extension module created from the `PyModule_Create()`, keep the old behavior. cc @vstinner

**Comment by encukou** [comment]
The relevant check is done by the `def->m_size == -1`. Extensions that use multi-phase init must set a positive `m_size`.

**Comment by shihai1991** [comment]
> The relevant check is done by the `def->m_size == -1`. Extensions that use multi-phase init must set a non-negative `m_size`.

Hm. I agree with you. But the user of bpo-44050 got the error in v3.9.6. And I found some module created from `PyModule_Create()` use `m_size`. For example: https://github.com/python/cpython/blob/v3.9.6/Modules/_posixsubprocess.c#L986

**Comment by shihai1991** [comment]
This PR can be removed when all the extension modules convert to multi-phase init.

**Comment by encukou** [comment]
Why is it important to ask whether the module was created by `PyModule_Create`?

**Comment by encukou** [comment]
> This PR can be removed when all the extension modules convert to multi-phase init.

That includes third-party modules, so: this PR can be removed when the API for single-phase init is removed.

**Comment by shihai1991** [comment]
> Why is it important to ask whether the module was created by `PyModule_Create`?

Make sure only the single extension module shared between interpreters. The extension module from multi-phase init can be created seperately in subinterpreter, right?

**Comment by encukou** [comment]
All modules with non-negative [m_size](https://docs.python.org/3/c-api/module.html#c.PyModuleDef.m_size) shouldn't be shared across interpreters, regardless of how they're created.

**Comment by shihai1991** [comment]
> All modules with non-negative [m_size](https://docs.python.org/3/c-api/module.html#c.PyModuleDef.m_size) shouldn't be shared across interpreters, regardless of how they're created.

Make sense. Looks your idea would be better. Thanks:) I will update it soon.

**Comment by vstinner** [comment]
_PyImport_FixupExtensionObject() copies the namespace from the first instance of an extension, to newly created instances of the extension. It's better than using exactly the same module object with the same __dict__ dictionary object. In a perfect world, all extensions would use multi-phase init and so don't go through _PyImport_FixupExtensionObject().

**Review by vstinner — APPROVED** [review]
LGTM, but can you try to write an unit test?

Would it be possible to write an unit test for this? https://bugs.python.org/issue44050 is a regression. It would be better to not reintroduce it by mistake tomorrow.

Maybe using the ssl module: if if  it's already imported, skip the test. Otherwise, spawn a subinterpreter to import it, and then import it in the main interpreter, and compare if you get the same objects or not.

See https://github.com/wee-slack/wee-slack/issues/812#issuecomment-818887462 for a more complete example.

To make sure that the ssl module is not imported yet, you can run the test in a subprocess, eg. using assert_python_ok(), or subprocess.

**Comment by shihai1991** [comment]
> LGTM, but can you try to write an unit test?
> 
> Would it be possible to write an unit test for this? https://bugs.python.org/issue44050 is a regression. It would be better to not reintroduce it by mistake tomorrow.
> 
> Maybe using the ssl module: if if it's already imported, skip the test. Otherwise, spawn a subinterpreter to import it, and then import it in the main interpreter, and compare if you get the same objects or not.
> 
> See [wee-slack/wee-slack#812 (comment)](https://github.com/wee-slack/wee-slack/issues/812#issuecomment-818887462) for a more complete example.
> 
> To make sure that the ssl module is not imported yet, you can run the test in a subprocess, eg. using assert_python_ok(), or subprocess.

OK, victor. I will update it.

**Comment by encukou** [comment]
> Maybe using the ssl module

That won't do; ssl uses multi-phase init (as should all other stdlib modules, one day). This would need to be a new module dedicated for the test.
You can look at `Modules/_testmultiphase.c` and `Lib/test/test_capi.py` for an example of adding a new module without adding a new shared library.

**Comment by shihai1991** [comment]
> adding a new module without adding a new shared library.

Using a new shared library should be better, no?

**Comment by encukou** [comment]
> Using a new shared library should be better, no?

I don't think so. Adding a new shared library module is a heavy process, e.g. it would need to be added to build instructions (for Unix and Windows), and distributors of Python will need to decide if they want to include the new test-only library.
But adding it to an existing library might come with its own issues, too...

**Comment by shihai1991** [comment]
> I don't think so. Adding a new shared library module is a heavy process, e.g. it would need to be added to build instructions (for Unix and Windows)

Agree. I try to add a shared Library in Windows at first.  I haven't experience to build python in VSCode, so I give it up finally :(
Adding a single-phase init module in _testmultiphase.c  looks like not very exact. But I think is acceptable.

> and distributors of Python will need to decide if they want to include the new test-only library.

OK. I saw the spec of fedora, using a python-test rpm to include the test files.

**Comment by encukou** [comment]
Looks good, although it looks like it'll only be backported to 3.10.1.
Could you add a NEWS entry, though? We should have those for bugfixes.

**Comment by trygveaa** [comment]
> Looks good, although it looks like it'll only be backported to 3.10.1.

What do you mean? The regression is in 3.9, the issue is fixed in 3.10, and 3.10.1 doesn't exist yet. Did you mean 3.9.1?

**Comment by shihai1991** [comment]
> > Looks good, although it looks like it'll only be backported to 3.10.1.
> 
> What do you mean? The regression is in 3.9, the issue is fixed in 3.10, and 3.10.1 doesn't exist yet. Did you mean 3.9.1?

Maybe we cloud backport to 3.9 and 3.10?

File: Misc/NEWS.d/next/Core and Builtins/2021-09-08-00-30-09.bpo-44050.mFI15u.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Extensions and def->m_base.m_copy can be updated when the extension module
+doesn't support sub-interpreters.
```
Comment (encukou) [inline Misc/NEWS.d/next/Core and Builtins/2021-09-08-00-30-09.bpo-44050.mFI15u.rst]:
```suggestion
Extensions that indicate they use global state (by setting ``m_size`` to -1)
can again be used in multiple interpreters. This reverts to behavior of
Python 3.8.
```

**Comment by encukou** [comment]
> What do you mean? The regression is in 3.9, the issue is fixed in 3.10, and 3.10.1 doesn't exist yet. Did you mean 3.9.1?

The "fix" in 3.10 is in the `_ssl` module only – an unrelated change that made the module avoid this bug.

Backporting to 3.10.1 & the next 3.9 sounds fine.

**Review by encukou — APPROVED** [review]

**Pull request merged**